src/main/java/org/jabref/gui/actions/ExportLinkedFilesAction.java

```diff
@@ -0,0 +1,120 @@
+package org.jabref.gui.actions;
+
+import java.awt.event.ActionEvent;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import javax.swing.AbstractAction;
+
+import javafx.concurrent.Service;
+import javafx.concurrent.Task;
+import javafx.scene.control.Button;
+import javafx.scene.control.ButtonType;
+import javafx.scene.control.DialogPane;
+
+import org.jabref.Globals;
+import org.jabref.JabRefGUI;
+import org.jabref.gui.DialogService;
+import org.jabref.gui.FXDialogService;
+import org.jabref.gui.util.DefaultTaskExecutor;
+import org.jabref.gui.util.DirectoryDialogConfiguration;
+import org.jabref.logic.l10n.Localization;
+import org.jabref.logic.util.io.FileUtil;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.LinkedFile;
+import org.jabref.model.util.FileHelper;
+import org.jabref.model.util.OptionalUtil;
+import org.jabref.preferences.JabRefPreferences;
+
+import org.controlsfx.dialog.ProgressDialog;
+
+public class ExportLinkedFilesAction extends AbstractAction {
+
+    private final BiFunction<Path, Path, Path> resolvePathFilename = (p, f) -> {
+        return p.resolve(f.getFileName());
+    };
+    private final DialogService ds = new FXDialogService();
+
+    public ExportLinkedFilesAction() {
+        super(Localization.lang("Export linked files"));
+    }
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+
+        DirectoryDialogConfiguration dirDialogConfiguration = new DirectoryDialogConfiguration.Builder()
+                .withInitialDirectory(
+                        Paths.get(Globals.prefs.get(JabRefPreferences.EXPORT_WORKING_DIRECTORY)))
+                .build();
+        List<BibEntry> entries = JabRefGUI.getMainFrame().getCurrentBasePanel().getSelectedEntries();
+
+        Optional<Path> path = DefaultTaskExecutor
+                .runInJavaFXThread(() -> ds.showDirectorySelectionDialog(dirDialogConfiguration));
+
+        BibDatabaseContext databaseContext = JabRefGUI.getMainFrame().getCurrentBasePanel().getDatabaseContext();
+
+        long totalFilesCount = entries.stream().flatMap(entry -> entry.getFiles().stream()).count();
+
+        Service<Void> service = new Service<Void>() {
```

Comment:
The enclosing `actionPerformed()` is a really big method. Please do some refactoring and make it smaller. The first thing would be to extract this class, at least to an inner class instead of an anonymous one.